src/models/Function/FunctionService.ts

```diff
@@ -0,0 +1,89 @@
+        const funcAppObj = await requestUtil(queryUrl, credentials);
+        // tslint:disable-next-line: no-unsafe-any
+        return JSON.parse(funcAppObj).value;
+    }
+
+    // check if the imported function app already has the hostkey, otherwise created
+    export async function addFuncHostKey(baseUrl: string, credentials: ServiceClientCredentials, serviceName: string): Promise<string> {
+        const hostKeys = await getFuncHostKeys(baseUrl, credentials);
+        const funcAppKeyName = `apim-${serviceName}`;
+        if (hostKeys !== undefined && hostKeys.functionKeys !== undefined && hostKeys.functionKeys[funcAppKeyName]) {
+            return hostKeys.functionKeys[funcAppKeyName];
+        }
+        ext.outputChannel.appendLine(localize("importFunctionApp", `Can't find existing function app host key ${funcAppKeyName}`));
+        return await createFuncHostKey(baseUrl, funcAppKeyName, credentials);
+    }
+
+    // Get function hostkey
+    export async function getFuncHostKeys(baseUrl: string, credentials: ServiceClientCredentials): Promise<IFunctionKeys> {
+        const options = new WebResource();
+        options.method = "POST";
+        options.url = `${baseUrl}/host/default/listkeys?api-version=${Constants.functionAppApiVersion}`;
+        options.headers = {
+            ['User-Agent']: appendExtensionUserAgent()
+        };
+        await signRequest(options, credentials);
+        // tslint:disable-next-line: await-promise
+        const funcKeys = await request(options).promise();
+        // tslint:disable-next-line: no-unsafe-any
+        return JSON.parse(funcKeys);
+    }
+
+    // Create function hostkey for our apim service
+    export async function createFuncHostKey(baseUrl: string, funcKeyName: string, credentials: ServiceClientCredentials): Promise<string> {
+        ext.outputChannel.appendLine(localize("importFunctionApp", `Create new function app host key ${funcKeyName}`));
+        // create new webresource
+        const options = new WebResource();
+        options.method = "PUT";
+        options.url = `${baseUrl}/host/default/functionkeys/${funcKeyName}?api-version=2018-11-01`;
+        options.headers = {
+            ['User-Agent']: appendExtensionUserAgent()
+        };
+        await signRequest(options, credentials);
+        // send request to get new function hostkey
+        const funKeyRequest = <nRequest>options;
+        funKeyRequest.body = {
+            properties: {}
+        };
+        funKeyRequest.json = true;
+        const response = await sendRequest(funKeyRequest);
+        if (response.properties.value) {
+            return String(response.properties.value);
+        } else {
+            throw new Error("Unexpected null value when generating function host key");
+        }
+    }
+
+    // Create client for webiste mgmt client
+    export function getClient(node: ApisTreeItem | ApiTreeItem): WebSiteManagementClient {
```

Comment:
dont pass node but pass credentials, subscription and environemt; Make this a generic and move to the baseclass.